src/mrpro/operators/_FourierOp.py

```diff
@@ -0,0 +1,262 @@
+        oversampling: SpatialDimension[float] = SpatialDimension(
+            2.0, 2.0, 2.0
+        ),  # TODO: not the nicest, since only used for nuFFT
+        norm='ortho',
+    ):
+        grid_size = []
+        nufft_im_size = []
+        nufft_dims = []
+        fft_dims = []
+        ignore_dims = []
+        fft_s = []
+        omega = []
+        traj_shape = []
+        super().__init__()
+
+        for n, os, k, i in zip(
+            (im_shape.z, im_shape.y, im_shape.x),
+            (oversampling.z, oversampling.y, oversampling.x),
+            (traj.kz, traj.ky, traj.kx),
+            (-3, -2, -1),
+        ):
+            nk_list = [traj.kz.shape[i], traj.ky.shape[i], traj.kx.shape[i]]
+            if n <= 1 and nk_list.count(1) == 3:
+                # dimension with no Fourier transform
+                ignore_dims.append(i)
+
+            elif nk_list.count(1) == 2:  # and is_uniform(k): #TODO: maybe is_uniform never needed?
+                # dimension with FFT
+                nk = torch.tensor(nk_list)
+                supp = torch.where(nk != 1, nk, 0)
+                s = supp.max().item()
+
+                # append dimension and output shape for oversampled FFT
+                fft_dims.append(i)
+                fft_s.append(s)
+            else:
+                # dimension with nuFFT
+                grid_size.append(int(os * n))
+                nufft_im_size.append(n)
+                nufft_dims.append(i)
+
+                # TODO: can omega be created here already?
+                # seems necessary to do this in a different loop where we create a list
+                # of kz, ky and kx to be stacked if they are compatible
+                # omega.append(k.flatten(start_dim=-3))
+
+            traj_shape.append(k.shape)
+
+        if len(nufft_dims) != 0:  # if non-uniform directions were identified
+            for k in (traj.kz, traj.ky, traj.kx):
+                # check number of singleton dimensions
+                n_singleton_dims = k.shape[1:].count(1)
+                if n_singleton_dims <= 1:
+                    # if not is_uniform(k): #TODO: is_uniform maybe never needed?
+                    omega.append(k.flatten(start_dim=-3))
+
+            self._omega = torch.stack(omega, dim=-2)
+            self._fwd_nufft_op = KbNufft(im_size=nufft_im_size, grid_size=grid_size)
+            self._adj_nufft_op = KbNufftAdjoint(im_size=nufft_im_size, grid_size=grid_size)
+
+        self._ignore_dims = tuple(ignore_dims)
+        self._nufft_dims = tuple(nufft_dims)
+        self._fft_dims = tuple(fft_dims)
+        self._fft_s = tuple(fft_s)
+        self._kshape = torch.broadcast_shapes(*traj_shape)
+        self._im_shape = im_shape
+        self._nufft_im_size = nufft_im_size
+        self.norm = norm
+
+    @staticmethod
+    def get_target_pattern(fft_dims, nufft_dims, ignore_dims):
+        """Pattern to reshape image/k-space data to be able to perform nuFFT.
+
+        Parameters
```

Comment:
the fact that these are tuples should not be mentioned here but only in type hinting above

Reply:
done